client/network/src/service.rs

```diff
@@ -1281,637 +1346,599 @@ where
-				ServiceToWorkerMsg::RemoveSetReserved(protocol, peer_id) => this
-					.network_service
-					.behaviour_mut()
-					.user_protocol_mut()
-					.remove_set_reserved_peer(protocol, peer_id),
-				ServiceToWorkerMsg::AddKnownAddress(peer_id, addr) =>
-					this.network_service.behaviour_mut().add_known_address(peer_id, addr),
-				ServiceToWorkerMsg::AddToPeersSet(protocol, peer_id) => this
-					.network_service
-					.behaviour_mut()
-					.user_protocol_mut()
-					.add_to_peers_set(protocol, peer_id),
-				ServiceToWorkerMsg::RemoveFromPeersSet(protocol, peer_id) => this
-					.network_service
-					.behaviour_mut()
-					.user_protocol_mut()
-					.remove_from_peers_set(protocol, peer_id),
-				ServiceToWorkerMsg::EventStream(sender) => this.event_streams.push(sender),
-				ServiceToWorkerMsg::Request {
-					target,
-					protocol,
-					request,
-					pending_response,
-					connect,
-				} => {
-					this.network_service.behaviour_mut().send_request(
-						&target,
-						&protocol,
-						request,
-						pending_response,
-						connect,
-					);
-				},
-				ServiceToWorkerMsg::NetworkStatus { pending_response } => {
-					let _ = pending_response.send(Ok(this.status()));
-				},
-				ServiceToWorkerMsg::NetworkState { pending_response } => {
-					let _ = pending_response.send(Ok(this.network_state()));
-				},
-				ServiceToWorkerMsg::DisconnectPeer(who, protocol_name) => this
-					.network_service
-					.behaviour_mut()
-					.user_protocol_mut()
-					.disconnect_peer(&who, protocol_name),
-				ServiceToWorkerMsg::NewBestBlockImported(hash, number) => this
-					.network_service
-					.behaviour_mut()
-					.user_protocol_mut()
-					.new_best_block_imported(hash, number),
-			}
+	/// Perform one action on the network.
+	///
+	/// Returns `false` when the worker should be shutdown.
+	async fn next_action(&mut self) -> bool {
+		futures::select! {
+			// Next message from the service.
+			msg = self.from_service.next() => {
+				if let Some(msg) = msg {
+					self.handle_worker_message(msg);
+				} else {
+					return false
+				}
+			},
+			// Next event from `Swarm` (the stream guaranteed to never terminate).
+			event = self.network_service.select_next_some() => {
+				self.handle_swarm_event(event);
+			},
+		};
+
+		let num_connected_peers =
+			self.network_service.behaviour_mut().user_protocol_mut().num_connected_peers();
+
+		// Update the variables shared with the `NetworkService`.
+		self.num_connected.store(num_connected_peers, Ordering::Relaxed);
+		{
+			let external_addresses =
+				self.network_service.external_addresses().map(|r| &r.addr).cloned().collect();
+			*self.external_addresses.lock() = external_addresses;
+
+			let listen_addresses =
+				self.network_service.listeners().map(ToOwned::to_owned).collect();
+			*self.listen_addresses.lock() = listen_addresses;
+
+			let best_seen_block = self.best_seen_block();
+			*self.best_seen_block.lock() = best_seen_block;
 		}
```

Comment:
I have doubts that these values change every time we call `poll`.

Reply:
I'm also not sure why you use a `Mutex` here :shrug:

Reply:
And do we actually update the listen addresses after startup?

Reply:
(moved `best_seen_block` to `ChainSync`)

Reply:
> I tried making `NetworkService::listen_addresses` an async request to `NetworkWorker`, but this made `NetworkStateInfo` trait async and broke `impl<T> NetworkStateInfo for Arc<T>`.

I don't get this argument? You need to put the `async_trait` attribute above the impl and then it should work?

Reply:
For `impl<T> NetworkStateInfo for Arc<T>` to compile with `async fn listen_addresses(&self)` `T` must be `Sync + Send`. If that's OK to add these requirements to `NetworkService`,  I can make it async.